ravenml/data/helpers.py

```diff
@@ -0,0 +1,236 @@
+
+def or_filter(tags_df, filter_tags):
+    """Filters out a set of images based upon the union of its tag values
+    
+    NOTE: With an OR (union) filter, an image must possess at least one tag in
+    the `filter_tags` list in order to pass the filter.
+
+    Args:
+        tags_df (DataFrame): a pandas DataFrame storing image IDs and
+            associated tags; its structure is:
+                index (rows) = image ID (str)
+                column headers = the tags themselves
+                columns = True/False values for whether the image has the tag
+                    in that column header
+        filter_tags (list): a list of str values for the tags that should be
+            used to apply the filter
+    
+    Returns:
+        DataFrame: a subset of the input DataFrame with those images that
+            passed the OR filter remaining
+    """
+    result = pd.DataFrame()
+    for tag in filter_tags:
+        subset = tags_df[tags_df[tag]]
+        result = pd.concat((result, subset), sort=False)
+    result = result.reset_index().drop_duplicates(
+        subset='index', keep='first').set_index('index')
+    return result
+
+
+def join_sets(sets):
+    """Returns the union of a set of datasets
+    
+    NOTE: this function removes duplicates, so it is possible to end up with a
+    smaller number of items than are given as inputs.
+
+    Args:
+        sets (list): a list of datasets that should be merged into one (the
+            union of all datasets); each element of the list is a DataFrame
+            in the format:
+                index (rows) = image ID (str)
+                column headers = the tags themselves
+                columns = True/False values for whether the image has the tag
+                    in that column header
+    
+    Returns:
+        DataFrame: a DataFrame that represents the union of all of the input
+            DataFrames with duplicates removed; it stores image IDs and 
+            associated tags in the format:
+                index (rows) = image ID (str)
+                column headers = the tags themselves
+                columns = True/False values for whether the image has the tag
+                    in that column header
+    """
+    result = pd.DataFrame()
+    for group in sets:
+        result = pd.concat((result, group), sort=False)
+    result = result.reset_index().drop_duplicates(
+        subset='index', keep='first').set_index('index')
+    return result
+
+def ingest_metadata(imageset, bucket_name, cache):
+    only_json_func = lambda filename: filename.startswith("meta_")
+
+    # if data_source == "Local":
+    #     copy_data_locally(
+    #         source_dir=kwargs["data_filepath"], condition_func=only_json_func)
+    # elif data_source == "S3":
+
+    download_data_from_s3(
+        bucket_name=bucket_name,
+        filter_vals=imageset,
+        condition_func=only_json_func,
+        cache=cache)
+
+def load_metadata(METADATA_PREFIX, cache):
```

Comment:
again, name of function isn't logical. this is actually loading the tags, not the metadata. the metadata is only loaded as a means of returning a DataFrame of tags.